modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx

```diff
@@ -0,0 +1,485 @@
+}: IHeadlessResourceItemProps) => {
+	const fuzzyNameMatch = fuzzy.match(
+		query,
+		headlessResource.name,
+		FUZZY_OPTIONS
+	);
+
+	const fuzzyBundleLabelMatch = fuzzy.match(
+		query,
+		headlessResource.bundleLabel,
+		FUZZY_OPTIONS
+	);
+
+	return (
+		<ClayLayout.ContentRow className="headless-resource">
+			{fuzzyNameMatch ? (
+				<span
+					dangerouslySetInnerHTML={{
+						__html: fuzzyNameMatch.rendered,
+					}}
+				/>
+			) : (
+				<span>{headlessResource.name}</span>
+			)}
+
+			<span className="context">
+				{fuzzyBundleLabelMatch ? (
+					<span
+						dangerouslySetInnerHTML={{
+							__html: fuzzyBundleLabelMatch.rendered,
+						}}
+					/>
+				) : (
+					<span>{headlessResource.bundleLabel}</span>
+				)}
+
+				{` ${headlessResource.version}`}
+			</span>
+		</ClayLayout.ContentRow>
+	);
+};
+
+const RequiredMark = () => (
+	<>
+		<span className="inline-item-after reference-mark text-warning">
+			<ClayIcon symbol="asterisk" />
+		</span>
+		<span className="hide-accessible sr-only">
+			{Liferay.Language.get('required')}
+		</span>
+	</>
+);
+
+interface IDropdownMenuProps {
+	headlessResources: Array<HeadlessResource>;
+	setHeadlessResourceValidationError: Function;
+	setSelectedHeadlessResource: Function;
+}
+
+const DropdownMenu = ({
+	headlessResources: initialHeadlessResources,
+	setHeadlessResourceValidationError,
+	setSelectedHeadlessResource,
+}: IDropdownMenuProps) => {
+	const [headlessResources, setHeadlessResources] = useState<
+		Array<HeadlessResource>
+	>(initialHeadlessResources || []);
+	const [query, setQuery] = useState('');
+
+	const onSearch = (query: string) => {
+		setQuery(query);
+
+		const regexp = new RegExp(query, 'i');
+
+		setHeadlessResources(
+			query
+				? initialHeadlessResources.filter(
+						({bundleLabel, name}: HeadlessResource) => {
+							return (
+								bundleLabel.match(regexp) || name.match(regexp)
+							);
+						}
+				  ) || []
+				: initialHeadlessResources
+		);
+	};
+
+	return (
+		<>
+			<ClayDropDown.Search
+				aria-label={Liferay.Language.get('search')}
+				onChange={onSearch}
+				value={query}
+			/>
+
+			<ClayDropDown.ItemList items={headlessResources} role="listbox">
+				{(item: HeadlessResource) => (
+					<ClayDropDown.Item
+						key={item.entityClassName}
+						onClick={() => {
+							setSelectedHeadlessResource(item);
+
+							setHeadlessResourceValidationError(false);
+						}}
+						roleItem="option"
+					>
+						<HeadlessResourceItem
+							headlessResource={item}
+							query={query}
+						/>
+					</ClayDropDown.Item>
+				)}
+			</ClayDropDown.ItemList>
+		</>
+	);
+};
+
+interface IFDSEntriesProps {
+	apiURL: string;
+	headlessResources: Array<HeadlessResource>;
+	namespace: string;
+}
+
+const FDSEntries = ({
+	apiURL,
+	headlessResources,
+	namespace,
+}: IFDSEntriesProps) => {
+	const headlessResourcesMapRef = useRef<Map<string, HeadlessResource>>();
+
+	useEffect(() => {
+		headlessResourcesMapRef.current = new Map(
```

Comment:
A minor nit, I _think_ this effect is not needed, we can just compute the ref in the beginning as nobody is changing the headlessResources array between 2 renders.